### PR TITLE
Reduce one top perimeter offset

### LINKED ISSUE #3313: Perimeters offset too much for one wall on top surfaces
For lower top surfaces perimeters are shifted inward, especially when there are number of walls:

image

I guess instead of complex calculations here
https://github.com/SoftFever/OrcaSlicer/blob/7ab2fabe6a8b314facb05a0cea372f5b304dc63a/src/libslic3r/PerimeterGenerator.cpp#L885

something like `coord_t offset_top_surface = ext_perimeter_width;` could produce better result:

image

providing base for next layer external perimeter. Picture is from Prusa with ported code.

[Orca-Cube.zip](https://github.com/SoftFever/OrcaSlicer/files/13778968/Orca-Cube.zip)

### LINKED ISSUE #6000: One wall on top causes wrong wall generation in Arachne
### Is there an existing issue for this problem?

- [X] I have searched the existing issues

### OrcaSlicer Version

2.1.1

### Operating System (OS)

Windows

### OS Version

Win10x64

### Additional system information

_No response_

### Printer

Custom

### How to reproduce

Slice attached project with Arachne and one wall on top surfaces enabled.

### Actual results

Invalid inernal wall
image

### Expected results

Round wall generated (screenshot with one wall on top disabled).

image

### Project file & Debug log uploads

[round.zip](https://github.com/user-attachments/files/16078332/round.zip)

### Checklist of files to include

- [ ] Log file
- [X] Project file

### Anything else?

I didn't succeed to find Arachne setting to prevent it.

### PR DESCRIPTION
**Pull request by vovodroid**
Reduce offset of internal perimeters inward for one top perimeter. Current value seems to be too large causing excessive shift and artifacts. Fixes  #6000 **One wall on top causes wrong wall generation in Arachne ** and  fixes #3313 **Perimeters offset too much for one wall on top surfaces ** 

Current value:
image
image

New value:
image
image

Both Arache and Classib are affected.

Obviously this  value could be configurable, though it seems this value is good enough.

### CODE REVIEW

**Comment by igiannakas** [comment]
I think I had experimented with this a while ago but I found that the offset was needed at least with my printer to “tuck in” the top solid infill under the walls of the next layer. What has your experience been? Admittedly it was before implementing the independent top and bottom overlap % though.

**Comment by vovodroid** [comment]
With suggested offset it's almost "tucked in"

image

Not prefect, but I guess it's better than having completely broken perimeter. As I mentioned this coefficient can be configurable, or rather offset formula could be improved in general. For now I prefer reduced offset.

**Comment by igiannakas** [comment]
Maybe 1x would cover the delta nicely? 

Need to try the PR with more complex models with circular tops and also some of the Voron parts like the stealth burner cover that didn’t like one wall top surface before…

**Comment by vovodroid** [comment]
> Maybe 1x would cover the delta nicely?

Yes, but it breaks the round model from screenshot [round.zip](https://github.com/user-attachments/files/16078332/round.zip)

**Comment by igiannakas** [comment]
> > Maybe 1x would cover the delta nicely?
> 
> Yes, but it still breaks the round model from screenshot [round.zip](https://github.com/user-attachments/files/16078332/round.zip)

How does prusas implementation behave? Maybe we can port that over if it’s better?

**Comment by vovodroid** [comment]
> I think they have rewritten the feature

I guess they based on my PR of potting from Orca))) Bamboo also works good.

**Comment by igiannakas** [comment]
OK I've done a bit of testing on this and I think it's ok to be a bit inset. By the way I've also seen this aberration with the feature enabled - 0.16LH - 0.25 first layer height, layer 199.

image

Have you ever seen this before? Project file below:

[one wall issue.3mf.zip](https://github.com/user-attachments/files/16287527/one.wall.issue.3mf.zip)

**Comment by igiannakas** [comment]
Unfortunately it is end up causing ever so slight perimeter gaps as the overlap of the top surface is affected by it :(

IMG_4140
IMG_4141
The prusa implementation looks a bit different - haven’t had time to dissect it but it may be worth doing so

**Comment by vovodroid** [comment]
Here is layer 199:

image

What's wrong here?


> I think they have rewritten the feature

Yes, I took a look on they code - it's different.

**Comment by igiannakas** [comment]
> Here is layer 199:
> 
> image
> 
> What's wrong here?
> 
> > I think they have rewritten the feature
> 
> Yes, I took a look on they code - it's different.

It’s generating solid infill instead of the walls. That messes with wall ordering algos like IoI

**Comment by vovodroid** [comment]
> It’s generating solid infill instead of the walls. That messes with wall ordering algos like IoI

I'll try to port code from Prusa.

**Comment by vovodroid** [comment]
@igiannakas 

Here is result of porting Prusa algorithm:

image

Your benchy:

image


As a side effect one top wall is always enabled. I hope it could be fixed)))

**Comment by igiannakas** [comment]
Looks good I think?

**Comment by vovodroid** [comment]
@igiannakas 
would you like to test https://github.com/vovodroid/OrcaSlicer/tree/one-top-wall-fix ?

Changes:
1) Port Arachne code from Prusa
2) Set top perimeter offset for Classic to 0.9. It seems producing good results.

**Comment by igiannakas** [comment]
Would you mind pushing the changes to this PR so I can cherry pick it into my Dev branch to test together with all the features I've pushed recently too? 

BTW this looks excellent!

**Comment by vovodroid** [comment]
> Would you mind pushing the changes to this PR

It's another branch with other meaning. I think you can create PR to yourself and merge to yours.

**Comment by igiannakas** [comment]
Ah yes indeed! Merged in the dev branch. Running a build now and will test. Doesnt this supersede this PR or am I misunderstanding it?

**Comment by igiannakas** [comment]
Ok tested - it works but it breaks the top/bottom solid infill - wall overlap for surfaces that are not the topmost/bottom most. It only changes overlap based on the infill/wall overlap value for any intermediate top surfaces.

**Comment by vovodroid** [comment]
Could you provide an example project?

**Comment by vovodroid** [comment]
> Doesnt this supersede this PR

Yes, it's new different code.

**Comment by igiannakas** [comment]
> Could you provide an example project?

Yeap 

How its supposed to work (exaggerated for effect)
image

vs

image
Only top most surface adjusted

Project file: [Overlap.3mf.zip](https://github.com/user-attachments/files/16339066/Overlap.3mf.zip)

**Comment by igiannakas** [comment]
I think its because this piece of code is missing:

```
        if (!top_fills.empty()) {
            infill_exp = union_ex(infill_exp, offset_ex(top_fills, double(top_inset)));
        }
```

But without delving deeper in the code I'm not sure where the new top surface polygons are now held

**Comment by vovodroid** [comment]
Yes, I see.

> because this piece of code is missing:

`top_fills` was filled in deleted call to `split_surfaces`, I'll restore it for this purpose and test.

**Comment by vovodroid** [comment]
Ok, I found top poligons in Prusa code

image

New commit added to the same branch. If it seems to be good I'll create new PR.

**Comment by igiannakas** [comment]
Awesome thank you!! I’ll compile it shortly and let you know but the change seems correct so don’t see a reason it wouldn’t work :)

**Comment by igiannakas** [comment]
Tested - looks good! 

One last thing, I dont believe the one wall threshold parameter works any more with the prusa code. 

image

The below code is not evaluated (replaced with the prusa code)

```
    // skip if the exposed area is smaller than "min_width_top_surface"
    double min_width_top_surface = std::max(double(ext_perimeter_spacing / 2. + 10), scale_(config->min_width_top_surface.get_abs_value(unscale_(perimeter_width))));

    // get the Polygons upper the polygon this layer
    Polygons upper_polygons_series_clipped = ClipperUtils::clip_clipper_polygons_with_subject_bbox(*this->upper_slices, last_box);
    upper_polygons_series_clipped          = offset(upper_polygons_series_clipped, min_width_top_surface);
```

[...]

```
// get the not-top surface, from the "real top" but enlarged by external_infill_margin (and the
    // min_width_top_surface we removed a bit before)
    ExPolygons temp_gap = diff_ex(top_polygons, fill_clip);
    ExPolygons inner_polygons =
        diff_ex(orig_polygons,
                offset_ex(top_polygons, offset_top_surface + min_width_top_surface - double(ext_perimeter_spacing / 2.)),
                ApplySafetyOffset::Yes);
```

There is a question whether this is needed any more as I've noticed that this approach is far less sensitive to errors with text embossed on top surfaces 
(left old, right new)
image

However I did notice that Prusa's approach will result in over extrusion over thin wall surfaces, like letters

image

Notice that the outline is drawn with an internal perimeter but the top surface doesnt segment around it. **This will result in a nozzle collision over the letter.**

image

Prusa slicer has the exact same problem:

image

Whereas with the old version there were artefacts (tiny holes) but no nozzle collisions.
(0% one wall threshold)
image
(300% one wall threshold)
image

I suspect this is happening due to the difference in shrinking/expanding the polygons between the prusa approach and the old Orca approach.

Project file: [bbb.3mf.zip](https://github.com/user-attachments/files/16349587/bbb.3mf.zip)

However the prusa approach (and the port to orca) fixes another long standing issue, where the first internal perimeter when a one wall top surface was printed, had the wrong line width. See below (left old - right new) - look at the thickness of the first wall.

image

So in summary it's almost perfect, if only we could figure out the segmentation for the lettering issue. Everything else seems to be working better than the current implementation.

**Comment by vovodroid** [comment]
Does threshold value prevent nozzle collision in current version? I could try to use it.

**Comment by igiannakas** [comment]
> Does threshold value prevent nozzle collision in current version? I could try to use it.

I think so but I’m not 100% sure

**Comment by vovodroid** [comment]
New PR  One top wall fix pr #6236  open.
